crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -2431,6 +2444,16 @@ where
                 if self.settings.rules.enabled(Rule::BannedApi) {
                     flake8_tidy_imports::banned_api::banned_attribute_access(self, expr);
                 }
+                if self
+                    .settings
+                    .rules
+                    .enabled(Rule::MissingFutureAnnotationsWithImports)
+                    && analyze::typing::is_pep585_builtin(expr, &self.ctx)
```

Comment:
These should only be activated for minimum-version < Python 3.9, right? Since PEP 585 was part of Python 3.9 (so future annotations aren't required to use the standard-library variants in 3.9 and later).

Reply:
In practice, this works on Python 3.7+.

Reply:
But if your minimum-supported Python version is Python 3.9, you don't need `__future__` annotations to use the standard-library generics. So these errors would already be detected and fixed by the existing `pyupgrade` rules, right?

Reply:
Ah, you're asking about the upper version. Unions like `str | None` didn't start working until Python 3.10+, which are also covered by this plugin. After 3.10 the future annotations import is not as useful anymore.